File: src-tauri/src/database/time_record.rs
use std::collections::HashMap;
use rusqlite::{Connection, params, Result, Row};
use serde::{Serialize, Deserialize};
use super::operation::insert_operation;


#[derive(Serialize, Deserialize, Debug)]
pub struct TimeRecord {
    pub id: i64,
    pub date: String,
    pub cost: i64,
    pub content: String,
    pub event_type: String,
    pub time_type: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct TimeRecordGroup {
    pub date: String,
    pub time_records: Vec<TimeRecord>,
}

pub fn get_time_record_from_query_result(row: &Row) -> TimeRecord {
    TimeRecord {
        id: row.get(0).unwrap(),
        date: row.get(1).unwrap(),
        cost: row.get(2).unwrap(),
        content: row.get(3).unwrap(),
        event_type: row.get(4).unwrap(),
        time_type: row.get(5).unwrap(),
    }
}

pub fn init_time_record_table(conn: &Connection) -> Result<()> {
    conn.execute(
        "CREATE TABLE IF NOT EXISTS time_records (
            id INTEGER PRIMARY KEY,
            date TEXT NOT NULL,
            cost INTEGER NOT NULL,
            content TEXT,
            event_type TEXT NOT NULL,
            time_type TEXT NOT NULL
        )",
        [],
    )?;
    Ok(())
}

pub fn create_time_record(
    conn: &Connection,
    date: &str,
    cost: i64,
    content: &str,
    event_type: &str,
    time_type: &str,
) -> Result<TimeRecord> {
    let mut stmt = conn.prepare("INSERT INTO time_records (date, cost, content, event_type, time_type) VALUES (?1, ?2, ?3, ?4, ?5)")?;
    let result = stmt.insert(params![date, cost, content, event_type, time_type])?;

    match insert_operation(conn, result as i64, "time_record".to_string(), "insert".to_string()) {
        Ok(_) => {}
        Err(e) => {
            println!("插入操作记录错误: {}", e);
        }
    };

    // 在查询一次
    let mut stmt = conn.prepare("SELECT id, date, cost, content, event_type, time_type FROM time_records WHERE id = ?1")?;
    let mut rows = stmt.query(params![result])?;
    let row = rows.next()?.unwrap();
    Ok(get_time_record_from_query_result(&row))
}

pub fn update_time_record(
    conn: &Connection,
    id: i64,
    date: &str,
    cost: i64,
    content: &str,
    event_type: &str,
    time_type: &str,
) -> Result<TimeRecord> {
    let mut stmt = conn.prepare("UPDATE time_records SET date = ?1, cost = ?2, content = ?3, event_type = ?4, time_type = ?5 WHERE id = ?6")?;
    let result = stmt.execute(params![date, cost, content, event_type, time_type, id])?;

    match insert_operation(conn, result as i64, "time_record".to_string(), "update".to_string()) {
        Ok(_) => {}
        Err(e) => {
            println!("插入操作记录错误: {}", e);
        }
    };

    // 在查询一次
    let mut stmt = conn.prepare("SELECT id, date, cost, content, event_type, time_type FROM time_records WHERE id = ?1")?;
    let mut rows = stmt.query(params![id])?;
    let row = rows.next()?.unwrap();
    Ok(get_time_record_from_query_result(&row))
}

pub fn delete_time_record(
    conn: &Connection,
    id: i64,
) -> Result<usize> {
    let mut stmt = conn.prepare("DELETE FROM time_records WHERE id = ?1")?;
    let result = stmt.execute(params![id])?;
    Ok(result)
}

pub fn get_time_record_by_id(
    conn: &Connection,
    id: i64,
) -> Result<TimeRecord> {
    let mut stmt = conn.prepare("SELECT id, date, cost, content, event_type, time_type FROM time_records WHERE id = ?1")?;
    let mut rows = stmt.query(params![id])?;
    let row = rows.next()?.unwrap();
    Ok(get_time_record_from_query_result(&row))
}

pub fn get_all_time_records(
    conn: &Connection,
) -> Result<Vec<TimeRecordGroup>> {
    let mut stmt = conn.prepare("SELECT id, date, cost, content, event_type, time_type FROM time_records ORDER BY date DESC")?;
    let mut rows = stmt.query(params![])?;
    let mut res = Vec::new();
    let mut map = HashMap::new();
    while let Some(row) = rows.next()? {
        let time_record = get_time_record_from_query_result(&row);
        let date = time_record.date.clone();
        if map.contains_key(&date) {
            let time_records: &mut Vec<TimeRecord> = map.get_mut(&date).unwrap();
            time_records.push(time_record);
        } else {
            let time_records = vec![time_record];
            map.insert(date, time_records);
        }
    }
    for (date, time_records) in map {
        res.push(TimeRecordGroup {
            date,
            time_records,
        });
    }
    Ok(res)
}

pub fn get_time_records_by_date(
    conn: &Connection,
    date: &str,
) -> Result<Vec<TimeRecord>> {
    let mut stmt = conn.prepare("SELECT id, date, cost, content, event_type, time_type FROM time_records WHERE date = ?1 ORDER BY date DESC")?;
    let mut rows = stmt.query(params![date])?;
    let mut res = Vec::new();
    while let Some(row) = rows.next()? {
        res.push(get_time_record_from_query_result(&row));
    }
    Ok(res)
}



// 按照日期分组
pub fn get_time_records_by_time_range(
    conn: &Connection,
    start_date: &str,
    end_date: &str,
) -> Result<Vec<TimeRecordGroup>> {
    let mut stmt = conn.prepare("SELECT id, date, cost, content, event_type, time_type FROM time_records WHERE date >= ?1 AND date <= ?2 ORDER BY date DESC")?;
    let mut rows = stmt.query(params![start_date, end_date])?;
    let mut res = Vec::new();
    let mut map = HashMap::new();
    while let Some(row) = rows.next()? {
        let time_record = get_time_record_from_query_result(&row);
        let date = time_record.date.clone();
        if map.contains_key(&date) {
            let time_records: &mut Vec<TimeRecord> = map.get_mut(&date).unwrap();
            time_records.push(time_record);
        } else {
            let time_records = vec![time_record];
            map.insert(date, time_records);
        }
    }
    for (date, time_records) in map {
        res.push(TimeRecordGroup {
            date,
            time_records,
        });
    }
    Ok(res)
}

// 获取所有的事件类型，不能重复
pub fn get_all_event_types(
    conn: &Connection,
) -> Result<Vec<String>> {
    let mut stmt = conn.prepare("SELECT DISTINCT event_type FROM time_records")?;
    let mut rows = stmt.query(params![])?;
    let mut res = Vec::new();
    // 首先判断已经存在，如果存在或者为空就不添加
    while let Some(row) = rows.next()? {
        let event_type: String = row.get(0)?;
        // 判断 res 是否有 event_type
        if !res.contains(&event_type) && event_type != "" {
            res.push(event_type);
        }
    }
    Ok(res)
}

// 获取所有的时间类型，不能重复
pub fn get_all_time_types(
    conn: &Connection,
) -> Result<Vec<String>> {
    let mut stmt = conn.prepare("SELECT DISTINCT time_type FROM time_records")?;
    let mut rows = stmt.query(params![])?;
    let mut res = Vec::new();
    // 首先判断已经存在，如果存在或者为空就不添加
    while let Some(row) = rows.next()? {
        let time_type: String = row.get(0)?;
        // 判断 res 是否有 time_type
        if !res.contains(&time_type) && time_type != "" {
            res.push(time_type);
        }
    }
    Ok(res)
}

File: src-electron/main/modules/tables/time-record.ts
import Database from "better-sqlite3";
import { ITimeRecord, TimeRecordGroup } from "@/types";
import Operation from "./operation";

export default class TimeRecordTable {
  static initTable(db: Database.Database) {
    db.exec(`
      CREATE TABLE IF NOT EXISTS time_records (
        id INTEGER PRIMARY KEY NOT NULL,
        date TEXT NOT NULL,
        cost REAL NOT NULL,
        content TEXT NOT NULL,
        event_type TEXT NOT NULL,
        time_type TEXT NOT NULL
      )
    `);
  }

  static upgradeTable(_db: Database.Database) {
    // 暂无升级
  }

  static getListenEvents() {
    return {
      "create-time-record": this.createTimeRecord.bind(this),
      "update-time-record": this.updateTimeRecord.bind(this),
      "delete-time-record": this.deleteTimeRecord.bind(this),
      "get-time-record-by-id": this.getTimeRecordById.bind(this),
      "get-all-time-records": this.getAllTimeRecords.bind(this),
      "get-time-records-by-date": this.getTimeRecordsByDate.bind(this),
      "get-time-records-by-date-range":
        this.getTimeRecordsByDateRange.bind(this),
      "get-all-event-types": this.getAllEventTypes.bind(this),
      "get-all-time-types": this.getAllTimeTypes.bind(this),
    };
  }

  static parseTimeRecord(record: any): ITimeRecord {
    return {
      ...record,
      content: JSON.parse(record.content),
      eventType: record.event_type,
      timeType: record.time_type,
    };
  }

  static createTimeRecord(
    db: Database.Database,
    record: Omit<ITimeRecord, "id">,
  ): ITimeRecord {
    const stmt = db.prepare(`
      INSERT INTO time_records (date, cost, content, event_type, time_type)
      VALUES (?, ?, ?, ?, ?)
    `);
    const res = stmt.run(
      record.date,
      record.cost,
      JSON.stringify(record.content),
      record.eventType,
      record.timeType,
    );

    Operation.insertOperation(
      db,
      "time-record",
      "insert",
      res.lastInsertRowid,
      Date.now(),
    );

    return this.getTimeRecordById(db, Number(res.lastInsertRowid));
  }

  static updateTimeRecord(
    db: Database.Database,
    record: ITimeRecord,
  ): ITimeRecord {
    const stmt = db.prepare(`
      UPDATE time_records SET
        date = ?,
        cost = ?,
        content = ?,
        event_type = ?,
        time_type = ?
      WHERE id = ?
    `);
    stmt.run(
      record.date,
      record.cost,
      JSON.stringify(record.content),
      record.eventType,
      record.timeType,
      record.id,
    );

    Operation.insertOperation(
      db,
      "time_record",
      "update",
      record.id,
      Date.now(),
    );

    return this.getTimeRecordById(db, record.id);
  }

  static deleteTimeRecord(db: Database.Database, id: number): number {
    const stmt = db.prepare("DELETE FROM time_records WHERE id = ?");
    Operation.insertOperation(db, "time_record", "delete", id, Date.now());
    return stmt.run(id).changes;
  }

  static getTimeRecordById(db: Database.Database, id: number): ITimeRecord {
    const stmt = db.prepare("SELECT * FROM time_records WHERE id = ?");
    const record = stmt.get(id);
    return this.parseTimeRecord(record);
  }

  static getTimeRecordsByDate(
    db: Database.Database,
    date: string,
  ): ITimeRecord[] {
    const stmt = db.prepare("SELECT * FROM time_records WHERE date = ?");
    const records = stmt.all(date);
    return records.map(this.parseTimeRecord);
  }

  static getTimeRecordsByDateRange(
    db: Database.Database,
    startDate: string,
    endDate: string,
  ): TimeRecordGroup {
    const stmt = db.prepare(`
      SELECT * FROM time_records 
      WHERE date BETWEEN ? AND ?
      ORDER BY date
    `);
    const records = stmt.all(startDate, endDate);

    const grouped: { [date: string]: ITimeRecord[] } = {};
    records.forEach((record) => {
      const parsed = this.parseTimeRecord(record);
      if (!grouped[parsed.date]) {
        grouped[parsed.date] = [];
      }
      grouped[parsed.date].push(parsed);
    });

    return Object.entries(grouped).map(([date, timeRecords]) => ({
      date,
      timeRecords,
    }));
  }

  static getAllTimeRecords(db: Database.Database): TimeRecordGroup {
    const stmt = db.prepare("SELECT * FROM time_records ORDER BY date");
    const records = stmt.all();

    const grouped: { [date: string]: ITimeRecord[] } = {};
    records.forEach((record) => {
      const parsed = this.parseTimeRecord(record);
      if (!grouped[parsed.date]) {
        grouped[parsed.date] = [];
      }
      grouped[parsed.date].push(parsed);
    });

    return Object.entries(grouped).map(([date, timeRecords]) => ({
      date,
      timeRecords,
    }));
  }

  static getAllEventTypes(db: Database.Database): string[] {
    const stmt = db.prepare("SELECT DISTINCT event_type FROM time_records");
    const types = stmt.all() as { event_type: string }[];
    return types.map((t) => t.event_type);
  }

  static getAllTimeTypes(db: Database.Database): string[] {
    const stmt = db.prepare("SELECT DISTINCT time_type FROM time_records");
    const types = stmt.all() as { time_type: string }[];
    return types.map((t) => t.time_type);
  }

  static getTimeRecordsByEventType(
    db: Database.Database,
    eventType: string,
  ): ITimeRecord[] {
    const stmt = db.prepare(`
      SELECT * FROM time_records 
      WHERE event_type = ?
      ORDER BY date
    `);
    const records = stmt.all(eventType);

    return records.map(this.parseTimeRecord);
  }

  static getTimeRecordsByTimeType(
    db: Database.Database,
    timeType: string,
  ): ITimeRecord[] {
    const stmt = db.prepare(`
      SELECT * FROM time_records 
      WHERE time_type = ?
      ORDER BY date
    `);
    const records = stmt.all(timeType);

    return records.map(this.parseTimeRecord);
  }

  static getTimeRecordsByEventTypeAndTimeType(
    db: Database.Database,
    eventType: string,
    timeType: string,
  ): ITimeRecord[] {
    const stmt = db.prepare(`
      SELECT * FROM time_records 
      WHERE event_type = ? AND time_type = ?
      ORDER BY date
    `);
    const records = stmt.all(eventType, timeType);

    return records.map(this.parseTimeRecord);
  }
}
